File: src/backend/app/schemas/files.py
from typing import Any, Literal

from pydantic import BaseModel


FileStatus = Literal["uploaded", "extracted", "failed", "deleted"]
ExtractionStatus = Literal["succeeded", "failed"]


class FileExtractionResponse(BaseModel):
    """文件解析结果响应。"""

    id: str
    uploaded_file_id: str
    extractor_name: str
    status: ExtractionStatus
    text_content: str | None = None
    metadata: dict[str, Any] = {}
    error_message: str | None = None
    created_at: str
    updated_at: str


class UploadedFileResponse(BaseModel):
    """上传文件元数据响应。"""

    id: str
    thread_id: str | None = None
    original_filename: str
    content_type: str
    file_size_bytes: int
    sha256: str
    status: FileStatus
    created_at: str
    updated_at: str
    extraction: FileExtractionResponse | None = None


class UploadedFileListResponse(BaseModel):
    """某个会话的上传文件列表。"""

    items: list[UploadedFileResponse]

File: src/frontend/src/api/files.ts
export interface FileExtraction {
  id: string;
  uploaded_file_id: string;
  extractor_name: string;
  status: "succeeded" | "failed";
  text_content: string | null;
  metadata: Record<string, unknown>;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface UploadedFile {
  id: string;
  thread_id: string | null;
  original_filename: string;
  content_type: string;
  file_size_bytes: number;
  sha256: string;
  status: "uploaded" | "extracted" | "failed" | "deleted";
  created_at: string;
  updated_at: string;
  extraction: FileExtraction | null;
}

export interface UploadedFileList {
  items: UploadedFile[];
}

// 文件上传只把原始文件交给后端解析。前端不再用 FileReader 直接读取，
// 这样 PDF/DOCX 和解析失败状态都能统一由后端记录。
export async function uploadChatFile(file: File, threadId: string): Promise<UploadedFile> {
  const form = new FormData();
  form.append("file", file);
  form.append("thread_id", threadId);

  const response = await fetch("/api/files", {
    method: "POST",
    body: form,
  });

  if (!response.ok) {
    throw new Error(`Upload file failed with ${response.status}`);
  }

  return response.json() as Promise<UploadedFile>;
}

export async function fetchThreadFiles(threadId: string): Promise<UploadedFileList> {
  const response = await fetch(`/api/files?thread_id=${encodeURIComponent(threadId)}`);

  if (!response.ok) {
    throw new Error(`Fetch files failed with ${response.status}`);
  }

  return response.json() as Promise<UploadedFileList>;
}
